## Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

## Importing Data

In [2]:
path = r'/Users/nataliawijaya/Documents/Bootcamp/Instacart Basket Analysis/02 Data'

In [3]:
path

'/Users/nataliawijaya/Documents/Bootcamp/Instacart Basket Analysis/02 Data'

In [4]:
# Importing 'orders_products_merged_updated.pkl'
df_ords_prods_merged = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [5]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge       price_range_loc  \
0         Soda        77              7     9.0   both  Middle-range product   
1         Soda        77              7     9.0   both  Middle-range product   
2         Soda        77              7     9.0   both  Middle-range product   
3         Soda        77              7     9.0   both  Middle-range product   
4         Soda        77              7     9.0   both  Middle-range product   

      busiest_day     busiest_days busiest_period_of_day  
0  Regularly busy   Regularly busy        Average orders  
1  Regularly busy  Least busy days        Average orders  
2  Regularly busy  Least busy days           Most orders  
3      Least busy  Least busy days        Average orders  
4      Least busy  Least busy days           Most orders

In [6]:
df_ords_prods_merged.shape

(32404859, 18)

In [7]:
# Create a subset up to 1 million rows (:1000000)
df = df_ords_prods_merged[:1000000]

In [8]:
df.shape

(1000000, 18)

In [9]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge       price_range_loc  \
0         Soda        77              7     9.0   both  Middle-range product   
1         Soda        77              7     9.0   both  Middle-range product   
2         Soda        77              7     9.0   both  Middle-range product   
3         Soda        77              7     9.0   both  Middle-range product   
4         Soda        77              7     9.0   both  Middle-range product   

      busiest_day     busiest_days busiest_period_of_day  
0  Regularly busy   Regularly busy        Average orders  
1  Regularly busy  Least busy days        Average orders  
2  Regularly busy  Least busy days           Most orders  
3      Least busy  Least busy days        Average orders  
4      Least busy  Least busy days           Most orders

## Grouping Data with Pandas

To recap, you should always use the groupby() function as part of a series of steps, namely, the following:
1. Split the data into groups based on some criteria.
2. Apply a function to each group separately.
3. Combine the results into a dataframe or alternative data structure or create a new column in the current dataframe.

In [10]:
df.groupby('product_name')

## Aggregating Data with agg()

Aggregation functions create summaries of selected columns and store these summary values in new columns. These summaries usually take the form of a descriptive statistic, for instance, a mean, median, maximum, or minimum (the very concept that you encountered in pivot tables in Excel). Aggregate columns have more uses beyond the immediate analysis.

### Performing a Single Aggregation

If you were to calculate the mean of the “order_number” column grouped by the “department_id” column, you could quickly compare the average number of orders per user across each Instacart department. This will involve two of the steps:
1. Split the data into groups based on “department_id.”
2. Apply the agg() function to each group to obtain the mean values for the “order_number” column.

In [11]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge       price_range_loc  \
0         Soda        77              7     9.0   both  Middle-range product   
1         Soda        77              7     9.0   both  Middle-range product   
2         Soda        77              7     9.0   both  Middle-range product   
3         Soda        77              7     9.0   both  Middle-range product   
4         Soda        77              7     9.0   both  Middle-range product   

      busiest_day     busiest_days busiest_period_of_day  
0  Regularly busy   Regularly busy        Average orders  
1  Regularly busy  Least busy days        Average orders  
2  Regularly busy  Least busy days           Most orders  
3      Least busy  Least busy days        Average orders  
4      Least busy  Least busy days           Most orders

In [12]:
df.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

There are some aggregations that can be conducted without use of the agg() function. For instance, the command above could be replaced with a command that uses the mean() function to achieve the same results:

In [13]:
# SQUARE BRACKETS (In general, stick to brackets when it comes to indexing!)
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

Rather than including the column name within square brackets, as in the example above, you could also set the column name off with a dot, for instance:

In [14]:
# DOT NOTATION 
df.groupby('department_id').order_number.mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

### Performing Multiple Aggregations

In [15]:
df.groupby('department_id').agg({'order_number' : ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

### Aggregating Data with transform()

- Creating a loyalty flag column in your ords_prods_merge dataframe. 
- You’ll do this by first using the transform() function, which will create a new column containing the maximum frequency of the “order_number” column, and then using the loc() function, which will create a second column containing a flag designating whether a customer is “loyal” or not.

Step:
1. Split the data into groups based on the “user_id” column.
2. Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
3. Create a new column, “max_order,” into which you’ll place the results of your aggregation.

Once this process is complete, you can use the “max_order” column to create another new column that assigns a loyalty flag to each customer according to the criteria (via the loc() function).

In [16]:
# In this case, all three steps can be included in a single line of code
df_ords_prods_merged['max_order'] = df_ords_prods_merged.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/kj/ff_yw10j6_zbkk839hkzgftr0000gn/T/ipykernel_9441/4171536106.py:2: FutureWarning: The provided callable <function amax at 0x110f23ce0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merged['max_order'] = df_ords_prods_merged.groupby(['user_id'])['order_number'].transform(np.max)


In [17]:
df_ords_prods_merged.head(25)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   
20                     8.0         196                 14          1   
21                     4.0         196                  9          0   
22                     3.0         196                  1          1   
23                     NaN         196                  4          0   
24                     7.0         196                  5          1   

   product_name  aisle_id  department_id  prices _merge       price_range_loc  \
0          Soda      

### Deriving Columns with loc()

- If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
- If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
- If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

In [18]:
df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] <=10, 'loyalty_flag'] = 'New customer'

In [19]:
df_ords_prods_merged.loc[(df_ords_prods_merged['max_order'] > 10) & (df_ords_prods_merged['max_order'] <= 40), 'loyalty_flag'] = 'Regular customer'

In [20]:
df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [21]:
df_ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [22]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge       price_range_loc  \
0         Soda        77              7     9.0   both  Middle-range product   
1         Soda        77              7     9.0   both  Middle-range product   
2         Soda        77              7     9.0   both  Middle-range product   
3         Soda        77              7     9.0   both  Middle-range product   
4         Soda        77              7     9.0   both  Middle-range product   

      busiest_day     busiest_days busiest_period_of_day  max_order  \
0  Regularly busy   Regularly busy        Average orders         10   
1  Regularly busy  Least busy days        Average orders         10   
2  Regularly busy  Least busy days           Most orders         10   
3      Least busy  Least busy days        Average orders         10   
4      Least busy  Least busy days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [23]:
df_ords_prods_merged[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

#### This brings to the end of the three-step process:
1. Started by grouping your data,
2. Followed up by applying an aggregation function,
3. Finished everything off by using the resulting aggregation to create a flag column.

## Exercise

#### In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [24]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge       price_range_loc  \
0         Soda        77              7     9.0   both  Middle-range product   
1         Soda        77              7     9.0   both  Middle-range product   
2         Soda        77              7     9.0   both  Middle-range product   
3         Soda        77              7     9.0   both  Middle-range product   
4         Soda        77              7     9.0   both  Middle-range product   

      busiest_day     busiest_days busiest_period_of_day  max_order  \
0  Regularly busy   Regularly busy        Average orders         10   
1  Regularly busy  Least busy days        Average orders         10   
2  Regularly busy  Least busy days           Most orders         10   
3      Least busy  Least busy days        Average orders         10   
4      Least busy  Least busy days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [25]:
df_ords_prods_merged.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#### Analyze the result. How do the results for the entire dataframe differ from those of the subset? 

In [26]:
df.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

The average result of order_number from the subset (1000000) is greater than the average result of entire dataframe (except the department_id: 17)

#### The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [27]:
df_ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [28]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge       price_range_loc  \
0         Soda        77              7     9.0   both  Middle-range product   
1         Soda        77              7     9.0   both  Middle-range product   
2         Soda        77              7     9.0   both  Middle-range product   
3         Soda        77              7     9.0   both  Middle-range product   
4         Soda        77              7     9.0   both  Middle-range product   

      busiest_day     busiest_days busiest_period_of_day  max_order  \
0  Regularly busy   Regularly busy        Average orders         10   
1  Regularly busy  Least busy days        Average orders         10   
2  Regularly busy  Least busy days           Most orders         10   
3      Least busy  Least busy days        Average orders         10   
4      Least busy  Least busy days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [29]:
df_ords_prods_merged.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

##### Observation:

1. It is interesting to see that new customers have a higher average purchase price than the other 2 customer categories and even loyal customers have a lower average purchase than the others
2. The maximum price here seems odd compared to the mean value itself. The value of 99.999 has skewed the distribution so further analysis for this value will be performed

In [30]:
df_ords_prods_merged['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

In [31]:
df_ords_prods_merged['prices'].sort_values(ascending = False)

29165979    99999.0
29165985    99999.0
29165519    99999.0
29165518    99999.0
29165517    99999.0
             ...   
7870751         1.0
7870750         1.0
7870749         1.0
7870748         1.0
30034999        1.0
Name: prices, Length: 32404859, dtype: float64

In [32]:
df_ords_prods_merged['prices'].drop_duplicates().sort_values(ascending = False)

29165516    99999.0
10030345    14900.0
18504754       25.0
25758883       24.9
28425432       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

Observation: The 99.999 and 14.900 prices look weird (that's a lot higher than the others)

In [33]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'].isin([99999.0,14900.0])]

order_id  user_id  order_number  orders_day_of_week  \
10030345    912404       17            12                   2   
10030346    603376       17            22                   6   
10030347   3264360      135             2                   2   
10030348    892534      135             3                   0   
10030349    229704      342             8                   1   
...            ...      ...           ...                 ...   
29166209   2249946   204099            29                   0   
29166210   2363282   204099            31                   0   
29166211   3181945   204395            13                   3   
29166212   2486215   205227             7                   3   
29166213   2232716   205227            10                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
10030345                 14                     5.0       21553   
10030346                 16                     4.0       21553   
10030347                 21                    13.0       21553   
10030348                  8                    12.0       21553   
10030349                 19                    30.0       21553   
...                     ...                     ...         ...   
29166209                  8                     4.0       33664   
29166210                  9                     2.0       33664   
29166211                 15                     8.0       33664   
29166212                 20                     4.0       33664   
29166213                 11                     4.0       33664   

          add_to_cart_order  reordered                      product_name  \
10030345                  5          0  Lowfat 2% Milkfat Cottage Cheese   
10030346                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030347                  6          0  Lowfat 2% Milkfat Cottage Cheese   
10030348                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030349                  9          0  Lowfat 2% Milkfat Cottage Cheese   
...                     ...        ...                               ...   
29166209                  1          0             2 % Reduced Fat  Milk   
29166210                  1          1             2 % Reduced Fat  Milk   
29166211                 25          0             2 % Reduced Fat  Milk   
29166212                  8          0             2 % Reduced Fat  Milk   
29166213                 23          1             2 % Reduced Fat  Milk   

          aisle_id  department_id   prices _merge     price_range_loc  \
10030345       108             16  14900.0   both  High-range-product   
10030346       108             16  14900.0   both  High-range-product   
10030347       108             16  14900.0   both  High-range-product   
10030348       108             16  14900.0   both  High-range-product   
10030349       108             16  14900.0   both  High-range-product   
...            ...            ...      ...    ...                 ...   
29166209        84             16  99999.0   both  High-range-product   
29166210        84             16  99999.0   both  High-range-product   
29166211        84             16  99999.0   both  High-range-product   
29166212        84             16  99999.0   both  High-range-product   
29166213        84             16  99999.0   both  High-range-product   

             busiest_day     busiest_days busiest_period_of_day  max_order  \
10030345  Regularly busy   Regularly busy           Most orders         40   
10030346  Regularly busy   Regularly busy           Most orders         40   
10030347  Regularly busy   Regularly busy        Average orders          4   
10030348     Busiest day     Busiest days        Average orders          4   
10030349  Regularly busy     Busiest days        Average orders         16   
...                  ...              ...                   ...        ...   
29166209     Busiest day     Busiest days        Average orders         39   
29166210  

In [34]:
df_high_prices = df_ords_prods_merged.loc[df_ords_prods_merged['prices'].isin([99999.0,14900.0])]

In [35]:
df_high_prices['prices'].value_counts(dropna = False)

prices
14900.0    4429
99999.0     698
Name: count, dtype: int64

In [36]:
df_high_prices.describe()

order_id        user_id  order_number  orders_day_of_week  \
count  5.127000e+03    5127.000000   5127.000000         5127.000000   
mean   1.708876e+06  106117.185684     15.302126            2.750536   
std    9.838956e+05   58594.982659     16.691849            2.010892   
min    4.300000e+01      17.000000      1.000000            0.000000   
25%    8.642320e+05   58136.000000      4.000000            1.000000   
50%    1.693799e+06  106021.000000      9.000000            3.000000   
75%    2.567244e+06  156389.000000     20.000000            5.000000   
max    3.421047e+06  206049.000000     99.000000            6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        5127.000000             4776.000000   5127.000000   
mean           13.473376               11.899079  23201.815682   
std             4.137326                9.068019   4153.742747   
min             0.000000                0.000000  21553.000000   
25%            10.000000                5.000000  21553.000000   
50%            13.000000                8.000000  21553.000000   
75%            16.000000               16.000000  21553.000000   
max            23.000000               30.000000  33664.000000   

       add_to_cart_order    reordered     aisle_id  department_id  \
count        5127.000000  5127.000000  5127.000000         5127.0   
mean            7.986347     0.613614   104.732592           16.0   
std             6.868889     0.486968     8.231346            0.0   
min             1.000000     0.000000    84.000000           16.0   
25%             3.000000     0.000000   108.000000           16.0   
50%             6.000000     1.000000   108.000000           16.0   
75%            11.000000     1.000000   108.000000           16.0   
max            72.000000     1.000000   108.000000           16.0   

             prices    max_order  
count   5127.000000  5127.000000  
mean   26485.547494    29.791691  
std    29186.636448    24.929207  
min    14900.000000     3.000000  
25%    14900.000000    12.000000  
50%    14900.000000    21.000000  
75%    14900.000000    41.000000  
max    99999.000000    99.000000

##### Observation:
1. Incredibly high prices (14.900 and 99.999) both from the department_id 16 with product_id 21553 and 33664
2. Next step: Find the product_name with product_id 21553 and 33664

In [37]:
df_ords_prods_merged.loc[df_ords_prods_merged['product_id'].isin([21553, 33664])]

order_id  user_id  order_number  orders_day_of_week  \
10030345    912404       17            12                   2   
10030346    603376       17            22                   6   
10030347   3264360      135             2                   2   
10030348    892534      135             3                   0   
10030349    229704      342             8                   1   
...            ...      ...           ...                 ...   
29166209   2249946   204099            29                   0   
29166210   2363282   204099            31                   0   
29166211   3181945   204395            13                   3   
29166212   2486215   205227             7                   3   
29166213   2232716   205227            10                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
10030345                 14                     5.0       21553   
10030346                 16                     4.0       21553   
10030347                 21                    13.0       21553   
10030348                  8                    12.0       21553   
10030349                 19                    30.0       21553   
...                     ...                     ...         ...   
29166209                  8                     4.0       33664   
29166210                  9                     2.0       33664   
29166211                 15                     8.0       33664   
29166212                 20                     4.0       33664   
29166213                 11                     4.0       33664   

          add_to_cart_order  reordered                      product_name  \
10030345                  5          0  Lowfat 2% Milkfat Cottage Cheese   
10030346                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030347                  6          0  Lowfat 2% Milkfat Cottage Cheese   
10030348                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030349                  9          0  Lowfat 2% Milkfat Cottage Cheese   
...                     ...        ...                               ...   
29166209                  1          0             2 % Reduced Fat  Milk   
29166210                  1          1             2 % Reduced Fat  Milk   
29166211                 25          0             2 % Reduced Fat  Milk   
29166212                  8          0             2 % Reduced Fat  Milk   
29166213                 23          1             2 % Reduced Fat  Milk   

          aisle_id  department_id   prices _merge     price_range_loc  \
10030345       108             16  14900.0   both  High-range-product   
10030346       108             16  14900.0   both  High-range-product   
10030347       108             16  14900.0   both  High-range-product   
10030348       108             16  14900.0   both  High-range-product   
10030349       108             16  14900.0   both  High-range-product   
...            ...            ...      ...    ...                 ...   
29166209        84             16  99999.0   both  High-range-product   
29166210        84             16  99999.0   both  High-range-product   
29166211        84             16  99999.0   both  High-range-product   
29166212        84             16  99999.0   both  High-range-product   
29166213        84             16  99999.0   both  High-range-product   

             busiest_day     busiest_days busiest_period_of_day  max_order  \
10030345  Regularly busy   Regularly busy           Most orders         40   
10030346  Regularly busy   Regularly busy           Most orders         40   
10030347  Regularly busy   Regularly busy        Average orders          4   
10030348     Busiest day     Busiest days        Average orders          4   
10030349  Regularly busy     Busiest days        Average orders         16   
...                  ...              ...                   ...        ...   
29166209     Busiest day     Busiest days        Average orders         39   
29166210  

##### Observation about finding:
1. The price of 14.900 comes from Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)
2. The price of 99.999 comes from 2 % Reduced Fat Milk (with product_id 33664)

In [38]:
# Confirm the finding number 1
# The price of 14.900 comes from Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)
df_ords_prods_merged.loc[df_ords_prods_merged['prices'].isin([14900.0])]

order_id  user_id  order_number  orders_day_of_week  \
10030345    912404       17            12                   2   
10030346    603376       17            22                   6   
10030347   3264360      135             2                   2   
10030348    892534      135             3                   0   
10030349    229704      342             8                   1   
...            ...      ...           ...                 ...   
10034769   3172853   205650            18                   1   
10034770   2504315   205818             3                   5   
10034771   1108388   205818             5                   4   
10034772   1916142   206049             1                   2   
10034773    379732   206049             4                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
10030345                 14                     5.0       21553   
10030346                 16                     4.0       21553   
10030347                 21                    13.0       21553   
10030348                  8                    12.0       21553   
10030349                 19                    30.0       21553   
...                     ...                     ...         ...   
10034769                  9                     7.0       21553   
10034770                 15                     3.0       21553   
10034771                  5                     1.0       21553   
10034772                 17                     NaN       21553   
10034773                 14                     5.0       21553   

          add_to_cart_order  reordered                      product_name  \
10030345                  5          0  Lowfat 2% Milkfat Cottage Cheese   
10030346                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030347                  6          0  Lowfat 2% Milkfat Cottage Cheese   
10030348                  3          1  Lowfat 2% Milkfat Cottage Cheese   
10030349                  9          0  Lowfat 2% Milkfat Cottage Cheese   
...                     ...        ...                               ...   
10034769                 17          1  Lowfat 2% Milkfat Cottage Cheese   
10034770                 13          0  Lowfat 2% Milkfat Cottage Cheese   
10034771                  5          1  Lowfat 2% Milkfat Cottage Cheese   
10034772                  2          0  Lowfat 2% Milkfat Cottage Cheese   
10034773                  4          1  Lowfat 2% Milkfat Cottage Cheese   

          aisle_id  department_id   prices _merge     price_range_loc  \
10030345       108             16  14900.0   both  High-range-product   
10030346       108             16  14900.0   both  High-range-product   
10030347       108             16  14900.0   both  High-range-product   
10030348       108             16  14900.0   both  High-range-product   
10030349       108             16  14900.0   both  High-range-product   
...            ...            ...      ...    ...                 ...   
10034769       108             16  14900.0   both  High-range-product   
10034770       108             16  14900.0   both  High-range-product   
10034771       108             16  14900.0   both  High-range-product   
10034772       108             16  14900.0   both  High-range-product   
10034773       108             16  14900.0   both  High-range-product   

             busiest_day     busiest_days busiest_period_of_day  max_order  \
10030345  Regularly busy   Regularly busy           Most orders         40   
10030346  Regularly busy   Regularly busy           Most orders         40   
10030347  Regularly busy   Regularly busy        Average orders          4   
10030348     Busiest day     Busiest days        Average orders          4   
10030349  Regularly busy     Busiest days        Average orders         16   
...                  ...              ...                   ...        ...   
10034769  Regularly busy     Busiest days           Most orders         25   
10034770  

##### Observation:
4429 values found with the price $14.900

In [39]:
df_high_price_14900 = df_ords_prods_merged.loc[df_ords_prods_merged['prices'].isin([14900.0])]

In [40]:
df_high_price_14900['product_name'].value_counts(dropna = False)

product_name
Lowfat 2% Milkfat Cottage Cheese    4429
Name: count, dtype: int64

##### Observation: 
The Lowfat 2% Milkfat Cottage Cheese product has value_counts = 4429 which has the exact same number as the product which has a price of 14.900, meaning that the entire price of 14.900 belongs to the entire  Lowfat 2% Milkfat Cottage Cheese product 

In [41]:
# Confirm the finding number 2
# The price of 99.999 comes from 2 % Reduced Fat Milk (with product_id 33664)
df_ords_prods_merged.loc[df_ords_prods_merged['prices'].isin([99999.0])]

order_id  user_id  order_number  orders_day_of_week  \
29165516    183964      873             3                   0   
29165517   1851256      873             4                   6   
29165518   1915696     1893             1                   5   
29165519   2763293     1893             2                   4   
29165520   2564805     1893             4                   1   
...            ...      ...           ...                 ...   
29166209   2249946   204099            29                   0   
29166210   2363282   204099            31                   0   
29166211   3181945   204395            13                   3   
29166212   2486215   205227             7                   3   
29166213   2232716   205227            10                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
29165516                 10                     7.0       33664   
29165517                 12                    13.0       33664   
29165518                 17                     NaN       33664   
29165519                 16                    13.0       33664   
29165520                 17                    30.0       33664   
...                     ...                     ...         ...   
29166209                  8                     4.0       33664   
29166210                  9                     2.0       33664   
29166211                 15                     8.0       33664   
29166212                 20                     4.0       33664   
29166213                 11                     4.0       33664   

          add_to_cart_order  reordered           product_name  aisle_id  \
29165516                 11          0  2 % Reduced Fat  Milk        84   
29165517                  8          1  2 % Reduced Fat  Milk        84   
29165518                 10          0  2 % Reduced Fat  Milk        84   
29165519                  6          1  2 % Reduced Fat  Milk        84   
29165520                  3          1  2 % Reduced Fat  Milk        84   
...                     ...        ...                    ...       ...   
29166209                  1          0  2 % Reduced Fat  Milk        84   
29166210                  1          1  2 % Reduced Fat  Milk        84   
29166211                 25          0  2 % Reduced Fat  Milk        84   
29166212                  8          0  2 % Reduced Fat  Milk        84   
29166213                 23          1  2 % Reduced Fat  Milk        84   

          department_id   prices _merge     price_range_loc     busiest_day  \
29165516             16  99999.0   both  High-range-product     Busiest day   
29165517             16  99999.0   both  High-range-product  Regularly busy   
29165518             16  99999.0   both  High-range-product  Regularly busy   
29165519             16  99999.0   both  High-range-product      Least busy   
29165520             16  99999.0   both  High-range-product  Regularly busy   
...                 ...      ...    ...                 ...             ...   
29166209             16  99999.0   both  High-range-product     Busiest day   
29166210             16  99999.0   both  High-range-product     Busiest day   
29166211             16  99999.0   both  High-range-product  Regularly busy   
29166212             16  99999.0   both  High-range-product  Regularly busy   
29166213             16  99999.0   both  High-range-product     Busiest day   

             busiest_days busiest_period_of_day  max_order      loyalty_flag  
29165516     Busiest days           Most orders          8      New customer  
29165517   Regularly busy           Most orders          8      New customer  
29165518   Regularly busy        Average orders          6      New customer  
29165519  Least busy days           Most orders          6      New customer  
29165520     Busiest days        Average orders          6      New customer  
...                   ...                   ...        ...               ...  
29166209     Busiest 

##### Observation:
698 values found with the price $99.999

In [42]:
df_high_price_99999 = df_ords_prods_merged.loc[df_ords_prods_merged['prices'].isin([99999.0])]

In [43]:
df_high_price_99999['product_name'].value_counts(dropna = False)

product_name
2 % Reduced Fat  Milk    698
Name: count, dtype: int64

##### Observation:
The 2 % Reduced Fat  Milk product has value_counts = 698 which has the exact same number as the product which has a price of 99.999, meaning that the entire price of 99.999 belongs to the entire 2 % Reduced Fat  Milk product

##### Next Step:
1. Create a new dataframe with department_id = 16 (df_dept_16)
2. Then another dataframe that excludes the high prices (df_dept_16_normal) --> using ~ to exclude the data

In [44]:
df_ords_prods_merged.loc[df_ords_prods_merged['department_id'] == 16]

order_id  user_id  order_number  orders_day_of_week  \
35791      2539329        1             1                   2   
35792      3241517      140            14                   4   
35793      1204424      293             1                   0   
35794      2748748      293             2                   1   
35795      1216787      293             5                   4   
...            ...      ...           ...                 ...   
32404819   1517645   191198            67                   6   
32404820   1104816   198258            14                   1   
32404822   1812848   191508             5                   3   
32404823    638659   196973            11                   2   
32404836   3004721   197459             2                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
35791                     8                     NaN       14084   
35792                    16                     7.0       14084   
35793                    20                     NaN       14084   
35794                    20                    15.0       14084   
35795                    18                    17.0       14084   
...                     ...                     ...         ...   
32404819                  8                     4.0        1561   
32404820                 20                     4.0        1561   
32404822                 10                    22.0        9305   
32404823                 15                    30.0        9305   
32404836                 20                     4.0       44495   

          add_to_cart_order  reordered  \
35791                     2          0   
35792                    25          0   
35793                     2          0   
35794                     2          1   
35795                    10          1   
...                     ...        ...   
32404819                  3          1   
32404820                  8          0   
32404822                 14          0   
32404823                 38          0   
32404836                  1          0   

                                               product_name  aisle_id  \
35791               Organic Unsweetened Vanilla Almond Milk        91   
35792               Organic Unsweetened Vanilla Almond Milk        91   
35793               Organic Unsweetened Vanilla Almond Milk        91   
35794               Organic Unsweetened Vanilla Almond Milk        91   
35795               Organic Unsweetened Vanilla Almond Milk        91   
...                                                     ...       ...   
32404819                                    Original Shreds        21   
32404820                                    Original Shreds        21   
32404822                          Torrential Tropical Punch        84   
32404823                          Torrential Tropical Punch        84   
32404836  Lowfat Cherry Lime Supernova Kefir Cultured Mi...       120   

          department_id  prices _merge       price_range_loc     busiest_day  \
35791                16    12.5   both  Middle-range product  Regularly busy   
35792                16    12.5   both  Middle-range product      Least busy   
35793                16    12.5   both  Middle-range product     Busiest day   
35794                16    12.5   both  Middle-range product  Regularly busy   
35795                16    12.5   both  Middle-range product      Least busy   
...                 ...     ...    ...                   ...             ...   
32404819             16     1.9   both     Low-range product  Regularly busy   
32404820             16     1.9   both     Low-range product  Regularly busy   
32404822             16    10.1   both  Middle-range product  Regularly busy   
32404823             16    10.1   both  Middle-range product  Regularly busy   
32404836             16    10.3   both  Middle-range product      Least busy   

             busiest_days busiest_period_of_day  max_order      loyalty_flag  
35

In [45]:
# Create a new dataframe with department_id = 16 (df_dept_16)
df_dept_16 = df_ords_prods_merged.loc[df_ords_prods_merged['department_id'] == 16]

In [46]:
df_dept_16.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
35791   2539329        1             1                   2                  8   
35792   3241517      140            14                   4                 16   
35793   1204424      293             1                   0                 20   
35794   2748748      293             2                   1                 20   
35795   1216787      293             5                   4                 18   

       days_since_prior_order  product_id  add_to_cart_order  reordered  \
35791                     NaN       14084                  2          0   
35792                     7.0       14084                 25          0   
35793                     NaN       14084                  2          0   
35794                    15.0       14084                  2          1   
35795                    17.0       14084                 10          1   

                                  product_name  aisle_id  department_id  \
35791  Organic Unsweetened Vanilla Almond Milk        91             16   
35792  Organic Unsweetened Vanilla Almond Milk        91             16   
35793  Organic Unsweetened Vanilla Almond Milk        91             16   
35794  Organic Unsweetened Vanilla Almond Milk        91             16   
35795  Organic Unsweetened Vanilla Almond Milk        91             16   

       prices _merge       price_range_loc     busiest_day     busiest_days  \
35791    12.5   both  Middle-range product  Regularly busy   Regularly busy   
35792    12.5   both  Middle-range product      Least busy  Least busy days   
35793    12.5   both  Middle-range product     Busiest day     Busiest days   
35794    12.5   both  Middle-range product  Regularly busy     Busiest days   
35795    12.5   both  Middle-range product      Least busy  Least busy days   

      busiest_period_of_day  max_order    loyalty_flag  
35791        Average orders         10    New customer  
35792           Most orders         76  Loyal customer  
35793        Average orders          8    New customer  
35794        Average orders          8    New customer  
35795        Average orders          8    New customer

In [47]:
df_dept_16['product_name'].value_counts(dropna = False)

product_name
Organic Whole Milk                                           137905
Organic Half & Half                                           76360
Half & Half                                                   69217
Organic Whole String Cheese                                   59676
Organic Unsweetened Almond Milk                               57895
                                                              ...  
Pumpkin Spice Almond Nog                                          1
Greek Blended Cherry Fat Free Yogurt                              1
Kefir Raspberry                                                   1
Cinnamon Rice Pudding Snacks                                      1
Lowfat Cherry Lime Supernova Kefir Cultured Milk Smoothie         1
Name: count, Length: 3444, dtype: int64

In [48]:
# Then another dataframe that excludes the high prices (df_dept_16_normal) --> using ~ to exclude the data
df_dept_16_normal = df_dept_16.loc[~df_dept_16['prices'].isin([99999, 14900])]

In [49]:
df_dept_16_normal.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
35791   2539329        1             1                   2                  8   
35792   3241517      140            14                   4                 16   
35793   1204424      293             1                   0                 20   
35794   2748748      293             2                   1                 20   
35795   1216787      293             5                   4                 18   

       days_since_prior_order  product_id  add_to_cart_order  reordered  \
35791                     NaN       14084                  2          0   
35792                     7.0       14084                 25          0   
35793                     NaN       14084                  2          0   
35794                    15.0       14084                  2          1   
35795                    17.0       14084                 10          1   

                                  product_name  aisle_id  department_id  \
35791  Organic Unsweetened Vanilla Almond Milk        91             16   
35792  Organic Unsweetened Vanilla Almond Milk        91             16   
35793  Organic Unsweetened Vanilla Almond Milk        91             16   
35794  Organic Unsweetened Vanilla Almond Milk        91             16   
35795  Organic Unsweetened Vanilla Almond Milk        91             16   

       prices _merge       price_range_loc     busiest_day     busiest_days  \
35791    12.5   both  Middle-range product  Regularly busy   Regularly busy   
35792    12.5   both  Middle-range product      Least busy  Least busy days   
35793    12.5   both  Middle-range product     Busiest day     Busiest days   
35794    12.5   both  Middle-range product  Regularly busy     Busiest days   
35795    12.5   both  Middle-range product      Least busy  Least busy days   

      busiest_period_of_day  max_order    loyalty_flag  
35791        Average orders         10    New customer  
35792           Most orders         76  Loyal customer  
35793        Average orders          8    New customer  
35794        Average orders          8    New customer  
35795        Average orders          8    New customer

In [50]:
df_dept_16_normal['prices'].describe()

count    5.393620e+06
mean     8.335056e+00
std      4.098207e+00
min      1.000000e+00
25%      4.800000e+00
50%      8.800000e+00
75%      1.190000e+01
max      1.500000e+01
Name: prices, dtype: float64

In [51]:
df_dept_16_normal['prices'].max()

15.0

##### Observation:
After excluding unusual values, the max price becomes more reasonable

##### Fixing the problem:
1. Finding the most closest value to replace 14.900 (Lowfat 2% Milkfat Cottage Cheese, product_id = 21553)
2. Finding the most closest value to replace 99.999 (2 % Reduced Fat Milk, product_id = 33664)

In [52]:
# First: Observation on Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)
# Create a filter for products that contain 'Lowfat', '2%', 'Milkfat', and Cottage Cheese' in the name, in df_dept_16_normal
cottage_cheese = df_dept_16_normal.query('product_name.str.contains("Lowfat") & product_name.str.contains("2%") & product_name.str.contains("Cottage Cheese") & product_name.str.contains("Milkfat")')

In [53]:
cottage_cheese['product_name'].value_counts(dropna = False)

product_name
Small Curd Lowfat 2% Milkfat Cottage Cheese                                2397
Cottage Doubles Peach, 2% Milkfat Lowfat Cottage Cheese & Peach Topping     350
Grade A Pasteurized 2% Milkfat Lowfat Cottage Cheese with Pineapple          32
Cottage Cheese 2% Milkfat Lowfat,Blueberry Topping                           14
Name: count, dtype: int64

In [54]:
cottage_cheese['prices'].describe()

count    2793.000000
mean        7.005979
std         1.829113
min         2.400000
25%         7.700000
50%         7.700000
75%         7.700000
max        11.800000
Name: prices, dtype: float64

##### Observation:
Now the mean, min and max prices for cottage_cheese are 7.0, 2.4, 11.8. Since this result makes more sense, the price of product ID 21553, named 'Lowfat 2% Milkfat Cottage Cheese' will be replaced from 14.900 and inputed by the mean value mentioned above (7.005979)

In [55]:
# Second: Observation on 2 % Reduced Fat Milk (with product_id 33664)
# Create a filter for products that contain '2%', Reduced Fat', and 'Milk' in the name, in df_dept_16_normal
reduced_fat_milk = df_dept_16_normal.query('product_name.str.contains("2%") & product_name.str.contains("Reduced Fat") & product_name.str.contains("Milk")')

In [56]:
reduced_fat_milk['product_name'].value_counts(dropna = False)

product_name
Organic Reduced Fat 2% Milk                                              47839
2% Reduced Fat Milk                                                      37091
Organic 2% Reduced Fat Milk                                              23244
Reduced Fat 2% Milk                                                      17320
2% Reduced Fat Organic Milk                                              16356
Organic Milk Reduced Fat, 2% Milkfat                                     12737
Dha Omega 3 Reduced Fat 2% Milk                                          10148
Milk, Reduced Fat, 2% Milkfat                                             7558
Lactose Free 2% Reduced Fat Milk                                          7114
Grassmilk 2% Reduced Fat Milk                                             5049
Reduced Fat 2% Chocolate Milk                                             3243
Reduced Fat 2% Lactose-Free Milk                                          3050
Reduced Fat Milk 2%                    

In [57]:
reduced_fat_milk['prices'].describe()

count    202238.000000
mean          9.086900
std           3.064869
min           1.200000
25%           7.900000
50%           9.200000
75%          11.400000
max          14.700000
Name: prices, dtype: float64

##### Observation:
Now the mean, min and max prices for reduced_fat_milk are 9.08, 1.2, 14.7. Since this result makes more sense, the price of product ID 33664, named '2 % Reduced Fat Milk' will be reduced from 99.999 to 9.99 (which is close to their average value)

##### Fixing the problem:
Replace the price '99999' with '9.99' and impute the value '14900' with the mean price from cottage_cheese

In [58]:
df_ords_prods_merged['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

In [59]:
df_ords_prods_merged['prices'] = df_ords_prods_merged['prices'].replace({99999: 9.99, 14900: cottage_cheese['prices'].mean()})

In [60]:
df_ords_prods_merged['prices'].describe()

count    3.240486e+07
mean     7.790934e+00
std      4.241496e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

##### Observation:
Now the mean, std, min and max results make more sense than the previous output

#### Determine the difference between the spending habits of the three types of customers you identified. Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [61]:
df_ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [62]:
df_ords_prods_merged.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773499  1.0  25.0
New customer      7.801156  1.0  25.0
Regular customer  7.798207  1.0  25.0

##### Observation:
1. Now the mean value becomes lower and max value is more reasonable (instead of 99999, now 25)
2. The order of results is still the same: new customer holds the highest purchase price followed by regular customer then loyal customer, but the difference now is small compared to the previous result 

#### The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
- If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
- If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [63]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge       price_range_loc  \
0         Soda        77              7     9.0   both  Middle-range product   
1         Soda        77              7     9.0   both  Middle-range product   
2         Soda        77              7     9.0   both  Middle-range product   
3         Soda        77              7     9.0   both  Middle-range product   
4         Soda        77              7     9.0   both  Middle-range product   

      busiest_day     busiest_days busiest_period_of_day  max_order  \
0  Regularly busy   Regularly busy        Average orders         10   
1  Regularly busy  Least busy days        Average orders         10   
2  Regularly busy  Least busy days           Most orders         10   
3      Least busy  Least busy days        Average orders         10   
4      Least busy  Least busy days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

##### Using transform() function
- Creating a spending_flag column in df_ords_prods_merged dataframe.
- First create a new column containing the average of the “prices” column by using transform() function, and then using the loc() function which will create a second column containing a flag designating whether a customer is high spender or low spender.

In [64]:
df_ords_prods_merged['average_price'] = df_ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/kj/ff_yw10j6_zbkk839hkzgftr0000gn/T/ipykernel_9441/1229806165.py:1: FutureWarning: The provided callable <function mean at 0x110f2ca40> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merged['average_price'] = df_ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)


In [65]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge       price_range_loc  \
0         Soda  ...              7     9.0    both  Middle-range product   
1         Soda  ...              7     9.0    both  Middle-range product   
2         Soda  ...              7     9.0    both  Middle-range product   
3         Soda  ...              7     9.0    both  Middle-range product   
4         Soda  ...              7     9.0    both  Middle-range product   

      busiest_day     busiest_days busiest_period_of_day max_order  \
0  Regularly busy   Regularly busy        Average orders        10   
1  Regularly busy  Least busy days        Average orders        10   
2  Regularly busy  Least busy days           Most orders        10   
3      Least busy  Least busy days        Average orders        10   
4      Least busy  Least busy days           Most orders        10   

   loyalty_flag average_price  
0  New customer      6.367797  
1  New customer      6.367797  
2  New customer      6.367797  
3  New customer      6.367797  
4  New customer      6.367797  

[5 rows x 21 columns]

In [66]:
# Rounding the average 
df_ords_prods_merged['average_price'] = df_ords_prods_merged['average_price'].round(2)

In [67]:
df_ords_prods_merged['average_price']

0           6.37
1           6.37
2           6.37
3           6.37
4           6.37
            ... 
32404854    6.91
32404855    6.91
32404856    7.63
32404857    7.63
32404858    7.68
Name: average_price, Length: 32404859, dtype: float64

In [68]:
# Creating a 'spending_flag' flag
df_ords_prods_merged.loc[df_ords_prods_merged['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [69]:
# Creating a 'spending_flag' flag
df_ords_prods_merged.loc[df_ords_prods_merged['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [70]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices  _merge       price_range_loc     busiest_day  \
0         Soda  ...     9.0    both  Middle-range product  Regularly busy   
1         Soda  ...     9.0    both  Middle-range product  Regularly busy   
2         Soda  ...     9.0    both  Middle-range product  Regularly busy   
3         Soda  ...     9.0    both  Middle-range product      Least busy   
4         Soda  ...     9.0    both  Middle-range product      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag  \
0   Regularly busy        Average orders        10  New customer   
1  Least busy days        Average orders        10  New customer   
2  Least busy days           Most orders        10  New customer   
3  Least busy days        Average orders        10  New customer   
4  Least busy days           Most orders        10  New customer   

   average_price spending_flag  
0           6.37   Low spender  
1           6.37   Low spender  
2           6.37   Low spender  
3           6.37   Low spender  
4           6.37   Low spender  

[5 rows x 22 columns]

In [71]:
# Checking the frequency
df_ords_prods_merged['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     32284501
High spender      120358
Name: count, dtype: int64

#### In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
- If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
- If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
- If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [72]:
# Using transform() function
df_ords_prods_merged['median_days_prior_orders'] = df_ords_prods_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/kj/ff_yw10j6_zbkk839hkzgftr0000gn/T/ipykernel_9441/3944351444.py:2: FutureWarning: The provided callable <function median at 0x112c6eb60> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df_ords_prods_merged['median_days_prior_orders'] = df_ords_prods_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [73]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge       price_range_loc     busiest_day  \
0         Soda  ...    both  Middle-range product  Regularly busy   
1         Soda  ...    both  Middle-range product  Regularly busy   
2         Soda  ...    both  Middle-range product  Regularly busy   
3         Soda  ...    both  Middle-range product      Least busy   
4         Soda  ...    both  Middle-range product      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag  \
0   Regularly busy        Average orders        10  New customer   
1  Least busy days        Average orders        10  New customer   
2  Least busy days           Most orders        10  New customer   
3  Least busy days        Average orders        10  New customer   
4  Least busy days           Most orders        10  New customer   

  average_price  spending_flag median_days_prior_orders  
0          6.37    Low spender                     20.5  
1          6.37    Low spender                     20.5  
2          6.37    Low spender                     20.5  
3          6.37    Low spender                     20.5  
4          6.37    Low spender                     20.5  

[5 rows x 23 columns]

In [74]:
# Creating a 'order_frequency_flag' flag
df_ords_prods_merged.loc[df_ords_prods_merged['median_days_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [75]:
# Creating a 'order_frequency_flag' flag
df_ords_prods_merged.loc[(df_ords_prods_merged['median_days_prior_orders'] > 10) & (df_ords_prods_merged['median_days_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [76]:
# Creating a 'order_frequency_flag' flag
df_ords_prods_merged.loc[df_ords_prods_merged['median_days_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [77]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...       price_range_loc     busiest_day     busiest_days  \
0         Soda  ...  Middle-range product  Regularly busy   Regularly busy   
1         Soda  ...  Middle-range product  Regularly busy  Least busy days   
2         Soda  ...  Middle-range product  Regularly busy  Least busy days   
3         Soda  ...  Middle-range product      Least busy  Least busy days   
4         Soda  ...  Middle-range product      Least busy  Least busy days   

  busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0        Average orders        10  New customer          6.37   Low spender   
1        Average orders        10  New customer          6.37   Low spender   
2           Most orders        10  New customer          6.37   Low spender   
3        Average orders        10  New customer          6.37   Low spender   
4           Most orders        10  New customer          6.37   Low spender   

   median_days_prior_orders   order_frequency_flag  
0                      20.5  Non-frequent customer  
1                      20.5  Non-frequent customer  
2                      20.5  Non-frequent customer  
3                      20.5  Non-frequent customer  
4                      20.5  Non-frequent customer  

[5 rows x 24 columns]

In [78]:
# Checking the frequency
df_ords_prods_merged['order_frequency_flag'].value_counts()

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
nan                             5
Name: count, dtype: int64

In [79]:
# Checking the nan values
df_ords_prods_merged[df_ords_prods_merged['order_frequency_flag'].isnull()]

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, product_id, add_to_cart_order, reordered, product_name, aisle_id, department_id, prices, _merge, price_range_loc, busiest_day, busiest_days, busiest_period_of_day, max_order, loyalty_flag, average_price, spending_flag, median_days_prior_orders, order_frequency_flag]
Index: []

[0 rows x 24 columns]

In [80]:
# Checking the nan values
df_ords_prods_merged[df_ords_prods_merged['median_days_prior_orders'].isnull()]

order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593                  5          0   
25880002                  1          0   

                                               product_name  ...  \
13645692                            Organic Red Bell Pepper  ...   
17251990                               Goat Cheese Crumbles  ...   
17622767                           California Veggie Burger  ...   
24138593                              Organic Peeled Garlic  ...   
25880002  4X Ultra Concentrated Natural Laundry Detergen...  ...   

               price_range_loc  busiest_day  busiest_days  \
13645692  Middle-range product  Busiest day  Busiest days   
17251990  Middle-range product  Busiest day  Busiest days   
17622767     Low-range product  Busiest day  Busiest days   
24138593  Middle-range product  Busiest day  Busiest days   
25880002     Low-range product  Busiest day  Busiest days   

         busiest_period_of_day max_order  loyalty_flag average_price  \
13645692        Average orders         1  New customer          7.42   
17251990        Average orders         1  New customer          7.42   
17622767        Average orders         1  New customer          7.42   
24138593        Average orders         1  New customer          7.42   
25880002        Average orders         1  New customer          7.42   

         spending_flag  median_days_prior_orders order_frequency_flag  
13645692   Low spender                       NaN                  nan  
17251990   Low spender                       NaN                  nan  
17622767   Low spender                       NaN                  nan  
24138593   Low spender                       NaN                  nan  
25880002   Low spender                       NaN                  nan  

[5 rows x 24 columns]

In [81]:
# Checking the nan values
df_ords_prods_merged[df_ords_prods_merged['median_days_prior_orders'].isnull()][['user_id', 'days_since_prior_order', 'median_days_prior_orders', 'order_frequency_flag']]

user_id  days_since_prior_order  median_days_prior_orders  \
13645692   159838                     NaN                       NaN   
17251990   159838                     NaN                       NaN   
17622767   159838                     NaN                       NaN   
24138593   159838                     NaN                       NaN   
25880002   159838                     NaN                       NaN   

         order_frequency_flag  
13645692                  nan  
17251990                  nan  
17622767                  nan  
24138593                  nan  
25880002                  nan

In [82]:
# Check user_id 159838
df_ords_prods_merged[df_ords_prods_merged['user_id'] == 159838]

order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593                  5          0   
25880002                  1          0   

                                               product_name  ...  \
13645692                            Organic Red Bell Pepper  ...   
17251990                               Goat Cheese Crumbles  ...   
17622767                           California Veggie Burger  ...   
24138593                              Organic Peeled Garlic  ...   
25880002  4X Ultra Concentrated Natural Laundry Detergen...  ...   

               price_range_loc  busiest_day  busiest_days  \
13645692  Middle-range product  Busiest day  Busiest days   
17251990  Middle-range product  Busiest day  Busiest days   
17622767     Low-range product  Busiest day  Busiest days   
24138593  Middle-range product  Busiest day  Busiest days   
25880002     Low-range product  Busiest day  Busiest days   

         busiest_period_of_day max_order  loyalty_flag average_price  \
13645692        Average orders         1  New customer          7.42   
17251990        Average orders         1  New customer          7.42   
17622767        Average orders         1  New customer          7.42   
24138593        Average orders         1  New customer          7.42   
25880002        Average orders         1  New customer          7.42   

         spending_flag  median_days_prior_orders order_frequency_flag  
13645692   Low spender                       NaN                  nan  
17251990   Low spender                       NaN                  nan  
17622767   Low spender                       NaN                  nan  
24138593   Low spender                       NaN                  nan  
25880002   Low spender                       NaN                  nan  

[5 rows x 24 columns]

##### Observation
user_id 159838 seems to be a new customer. Therefore, I leave the value of NaN as it is

In [83]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...       price_range_loc     busiest_day     busiest_days  \
0         Soda  ...  Middle-range product  Regularly busy   Regularly busy   
1         Soda  ...  Middle-range product  Regularly busy  Least busy days   
2         Soda  ...  Middle-range product  Regularly busy  Least busy days   
3         Soda  ...  Middle-range product      Least busy  Least busy days   
4         Soda  ...  Middle-range product      Least busy  Least busy days   

  busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0        Average orders        10  New customer          6.37   Low spender   
1        Average orders        10  New customer          6.37   Low spender   
2           Most orders        10  New customer          6.37   Low spender   
3        Average orders        10  New customer          6.37   Low spender   
4           Most orders        10  New customer          6.37   Low spender   

   median_days_prior_orders   order_frequency_flag  
0                      20.5  Non-frequent customer  
1                      20.5  Non-frequent customer  
2                      20.5  Non-frequent customer  
3                      20.5  Non-frequent customer  
4                      20.5  Non-frequent customer  

[5 rows x 24 columns]

In [84]:
df_ords_prods_merged.shape

(32404859, 24)

#### Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder

In [85]:
# Exporting df_ords_prods_merged dataframe as “orders_products_merged_updated_2.pkl” in “Prepared Data” folder
df_ords_prods_merged.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged_updated_2.pkl'))